psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1504 @@
+#pragma omp parallel for schedule(static, 1)    
+    for (size_t Q = 0; Q < grid->blocks().size(); Q++) {
+
+        size_t thread = 0;
+#ifdef _OPENMP
+        thread = omp_get_thread_num();
+#endif
+
+        std::shared_ptr<BlockOPoints> block = grid->blocks()[Q];
+        int nbf_block = block->local_nbf();
+        int npoints_block = block->npoints();
+
+        // compute values of each basis function at each point in this block
+        point_funcs[thread]->compute_functions(block);
+
+        // the values we just computed (max_points x max_functions)
+        SharedMatrix point_values = point_funcs[thread]->basis_values()["PHI"]; 
+         
+        std::vector<int> bf_map = block->functions_local_to_global();
+
+        // resize point_values buffer to size of this block
+        SharedMatrix point_values_trim = std::make_shared<Matrix>("DFTGrid PHI Buffer", npoints_block, nbf_block); // points x bf_block
+        for (size_t p = 0; p < npoints_block; p++) {
+            for(size_t k = 0; k < nbf_block; k++) {
+                point_values_trim->set(p,k,point_values->get(p,k));
+            }
+        }
+
+        SharedMatrix C_lmo_slice = submatrix_rows(C_lmo, bf_map); //bf_block x naocc
+        SharedMatrix C_pao_slice = submatrix_rows(C_pao, bf_map); //bf_block x npao
+
+        // value of mo at each point squared
+        C_lmo_slice = linalg::doublet(point_values_trim, C_lmo_slice, false, false); // points x naocc
+        C_pao_slice = linalg::doublet(point_values_trim, C_pao_slice, false, false); // points x npao
+
+        for (size_t p = 0; p < npoints_block; p++) {
+            for (size_t i = 0; i < naocc; ++i) {
+                C_lmo_slice->set(p, i, pow(C_lmo_slice->get(p, i), 2));
+            }
+            for (size_t u = 0; u < npao; ++u) {
+                C_pao_slice->set(p, u, pow(C_pao_slice->get(p, u), 2));
+            }
+        }
+
+        SharedMatrix C_lmo_slice_w = std::make_shared<Matrix>(C_lmo_slice); // points x naocc
+        for (size_t p = 0; p < npoints_block; p++) {
+            C_lmo_slice_w->scale_row(0, p, block->w()[p]);
+        }
+
+        DOI_ij_temps[thread]->add(linalg::doublet(C_lmo_slice_w, C_lmo_slice, true, false)); // naocc x naocc
+        DOI_iu_temps[thread]->add(linalg::doublet(C_lmo_slice_w, C_pao_slice, true, false)); // naocc x npao
+
+    }
+    timer_off("Integration");
+
+    DOI_ij = std::make_shared<Matrix>("(i,j) Differential Overlap Integrals", naocc, naocc);
+    DOI_iu = std::make_shared<Matrix>("(i,u) Differential Overlap Integrals", naocc, nbf);
+
+    for(size_t thread = 0; thread < nthread; thread++) {
+        DOI_ij->add(DOI_ij_temps[thread]);
+        DOI_iu->add(DOI_iu_temps[thread]);
+    }
+
+    for(size_t i = 0; i < naocc; i++) {
+        for(size_t j = 0; j < naocc; j++) {
+            DOI_ij->set(i,j,sqrt(DOI_ij->get(i,j)));
+        }
+        for(size_t u = 0; u < nbf; u++) {
+            DOI_iu->set(i,u,sqrt(DOI_iu->get(i,u)));
+        }
+    }
```

Comment:
I'm wondering whether you should just add a `sqrt` function to the `Matrix` class and simplify this as e.g.
```
DOI_ij->sqrt();
DOI_iu->sqrt();
```